### PR TITLE
NOS Engine-Rust API & Rust HALs

### PR DESCRIPTION
**Pull request by Seabass247**
### Background
***NOS3*** lives within its own vagrant box, consists of several middleware and software components, such as **COSMOS** for ground control, **42** for nominal data, and **NOS Engine** for simulated buses, including i2c, uart, and spi.   This is the first step to integrate KubOS into NOS3 as flight software alternative to cFS. 

ISIS AntS, Clyde 3g EPS, and Novatel OEM6 will soon be backed by NOS3 simulators via this work.
### What's Included
- `nosengine-rust` NOS Engine-Rust API; contains safe rust wrappers around the NOS Engine C Interface
     - NOTE:  nosengine libs must be installed before the crate can link to the library 
- Rust HALs; conditionally compiles (`cargo build --features nos3`) for connection to nosengine's simulated buses
     - `rust_uart` HAL for NOS uart connections
     - `rust_i2c` HAL for NOS i2c connections

### CODE REVIEW

File: apis/nosengine-rust/src/ffi/mod.rs [hunk 1]
```diff
@@ -0,0 +1,508 @@
...
+    NonException,
+    Unimplemented,
+    Unknown,
+    Count,
+}
+
+extern "C" {
+    fn NE_error() -> ErrorCode;
+    fn NE_error_string() -> *const c_char;
+    //    fn NE_exception_string() -> *const c_char;
```
Comment (plauche) [inline apis/nosengine-rust/src/ffi/mod.rs]: Can this commented out code get deleted?

File: hal/rust-hal/SimConfig.toml [hunk 2]
```diff
@@ -0,0 +1,36 @@
+# This file contains configuration options for the simulated NOSEngine HALs.
```
Comment (plauche) [inline hal/rust-hal/SimConfig.toml]: How likely are these values to change on a per-user/per-run basis?

File: hal/rust-hal/rust-uart/src/lib.rs [hunk 3]
```diff
@@ -120,3 +129,97 @@ impl Stream for SerialStream {
...
+            port: sync::Arc::new(sync::Mutex::new(port)),
+            min_timeout: Duration::from_millis(min_timeout as u64),
+        })
+    }
+}
+
+#[cfg(feature = "nos3")]
+impl Stream for SerialStream {
+    fn write(&self, data: &[u8]) -> UartResult<()> {
+        self.port.lock().unwrap().write(data);
```
Comment (plauche) [inline hal/rust-hal/rust-uart/src/lib.rs]: Do we want to risk a possible panic with these calls to `lock().unwrap()`, or should we be translating a failure to lock into some `UartError`?

File: apis/nosengine-rust/src/client/mod.rs [hunk 4]
```diff
@@ -0,0 +1,447 @@
...
+pub struct DataNode {
+    node_ptr: *mut ffi::DataNodeHandle,
+    bus_ptr: Arc<Bus>,
+    /// Name of this data node
+    pub name: String,
+}
+
+impl DataNode {
+    /// Creates a data node on the supplied bus. Data node names need to be unique, but no error will be
+    /// thrown if they are not unique! For example, the following code will result in an invalid
```
[the remark this answers is not in the document]
Comment (Seabass247) [reply apis/nosengine-rust/src/client/mod.rs]: across multiple processes AFAIK

File: apis/nosengine-rust/src/client/spi.rs [hunk 5]
```diff
@@ -0,0 +1,241 @@
...
+    spi_ptr: *mut spi::SPIHandle,
+    /// NOSEngine connection string
+    pub connection: &'a str,
+    /// Name of this bus to which this master is connected
+    pub bus: &'a str,
+}
+
+impl<'a> SPIMaster<'a> {
+    /// This function creates a new SPI master on the given bus. There can be only one: If you attempt to create
+    /// another master on the same bus, this function will return `None`.
```
[the remark this answers is not in the document]
Comment (Seabass247) [reply apis/nosengine-rust/src/client/spi.rs]: `SPIError::SPICreationError`

File: apis/nosengine-rust/src/client/spi.rs [hunk 6]
```diff
@@ -0,0 +1,241 @@
...
+    pub connection: &'a str,
+    /// Name of the bus to which this slave is connected
+    pub bus: &'a str,
+    /// Chip select number of this slave
+    pub cs: u8,
+}
+
+impl<'a> SPISlave<'a> {
+    /// Construct a new SPI slave. The given callback will be run every time the master reads
+    /// from or writes to this slave. If a slave with the given chip select already exists on
```
[the remark this answers is not in the document]
Comment (Seabass247) [reply apis/nosengine-rust/src/client/spi.rs]: `SPIError::SPICreationError`

File: apis/nosengine-rust/src/client/uart.rs [hunk 7]
```diff
@@ -0,0 +1,286 @@
...
+    /// let uart2 = UART::new("uart3", "tcp://localhost:12001", "uart2", 11).unwrap();
+    ///
+    /// uart1.write(&[1u8, 2, 3, 4]);
+    /// let result = uart2.read(100);
+    /// assert_eq!(result, &[1u8, 2, 3, 4]);
+    /// ```
+    pub fn read(&self, num_bytes: usize) -> Vec<u8> {
+        let mut buffer: Vec<u8> = Vec::with_capacity(num_bytes);
+        let len = uart::uart_read(self.uart_ptr, buffer.as_mut_ptr(), num_bytes);
+        assert!(len <= num_bytes);
```
[the remark this answers is not in the document]
Comment (Seabass247) [reply apis/nosengine-rust/src/client/uart.rs]: `uart::uart_read` will always return `len` which is always less than or equal to  `num_bytes` AKA the `len` arg for `uart::uart_read`

File: apis/nosengine-rust/src/ffi/i2c.rs [hunk 8]
```diff
@@ -0,0 +1,293 @@
...
+#[derive(Debug, Eq, PartialEq)]
+#[repr(C)]
+#[allow(missing_docs)]
+pub enum I2CDirection {
+    Write,
+    Read,
+}
+
+/// This enum represents a handle to an opaque C struct.
+pub enum I2CHandle {}
```
[the remark this answers is not in the document]
Comment (Seabass247) [reply apis/nosengine-rust/src/ffi/i2c.rs]: This was a choice made by the original author of the api.

File: hal/rust-hal/SimConfig.toml (hunk 2 above)
Comment (Seabass247) [reply hal/rust-hal/SimConfig.toml]: Very unlikely as of the way we use Kubos now in basic nos3.  AFAIK per-user could possibly change where they've made major changes to nos3 and sims for their mission.

File: apis/nosengine-rust/src/client/uart.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (Seabass247) [reply apis/nosengine-rust/src/client/uart.rs]: I guess I can remove the assertion; it should never panic as long the ffi function `uart::uart_read()` adheres to said behavior.

**Comment by plauche** [comment]
@Seabass247 overall this is looking solid. Can you go through and add copyright headers to the new contributed source files? You can do our standard copyright block and then a `Contributed by` line, like what was done in the comms-service source https://github.com/kubos/kubos/blob/master/libs/comms-service/src/config.rs.

**Pull request merged**